internal/incremental/snapshot.go

```diff
@@ -0,0 +1,569 @@
+			// If options affect emit, then we need to do complete emit per compiler options
+			// otherwise only the js or dts that needs to emitted because its different from previously emitted options
+			var pendingEmitKind FileEmitKind
+			if tsoptions.CompilerOptionsAffectEmit(oldProgram.snapshot.options, snapshot.options) {
+				pendingEmitKind = GetFileEmitKind(snapshot.options)
+			} else {
+				pendingEmitKind = getPendingEmitKindWithOptions(snapshot.options, oldProgram.snapshot.options)
+			}
+			if pendingEmitKind != FileEmitKindNone {
+				// Add all files to affectedFilesPendingEmit since emit changed
+				for _, file := range files {
+					// Add to affectedFilesPending emit only if not changed since any changed file will do full emit
+					if !snapshot.changedFilesSet.Has(file.Path()) {
+						snapshot.addFileToAffectedFilesPendingEmit(file.Path(), pendingEmitKind)
+					}
+				}
+				snapshot.buildInfoEmitPending = true
+			}
+		}
+		if len(snapshot.semanticDiagnosticsPerFile) != len(snapshot.fileInfos) &&
+			oldProgram.snapshot.checkPending != snapshot.checkPending {
+			snapshot.buildInfoEmitPending = true
+		}
+	}
+	return snapshot
+}
+
+func fileAffectsGlobalScope(file *ast.SourceFile) bool {
+	// if file contains anything that augments to global scope we need to build them as if
+	// they are global files as well as module
+	if core.Some(file.ModuleAugmentations, func(augmentation *ast.ModuleName) bool {
+		return ast.IsGlobalScopeAugmentation(augmentation.Parent)
+	}) {
+		return true
+	}
+
+	if ast.IsExternalOrCommonJSModule(file) || ast.IsJsonSourceFile(file) {
+		return false
+	}
+
+	// For script files that contains only ambient external modules, although they are not actually external module files,
+	// they can only be consumed via importing elements from them. Regular script files cannot consume them. Therefore,
+	// there are no point to rebuild all script files if these special files have changed. However, if any statement
+	// in the file is not ambient external module, we treat it as a regular script file.
+	return file.Statements != nil &&
+		file.Statements.Nodes != nil &&
+		core.Some(file.Statements.Nodes, func(stmt *ast.Node) bool {
+			return !ast.IsModuleWithStringLiteralName(stmt)
+		})
+}
+
+func addReferencedFilesFromSymbol(file *ast.SourceFile, referencedFiles *collections.Set[tspath.Path], symbol *ast.Symbol) {
+	if symbol == nil {
+		return
+	}
+	for _, declaration := range symbol.Declarations {
+		fileOfDecl := ast.GetSourceFileOfNode(declaration)
+		if fileOfDecl == nil {
+			continue
+		}
+		if file != fileOfDecl {
+			referencedFiles.Add(fileOfDecl.Path())
+		}
+	}
+}
+
+// Get the module source file and all augmenting files from the import name node from file
+func addReferencedFilesFromImportLiteral(file *ast.SourceFile, referencedFiles *collections.Set[tspath.Path], checker *checker.Checker, importName *ast.LiteralLikeNode) {
+	symbol := checker.GetSymbolAtLocation(importName)
```

Comment:
Pretty shocked that this is how it used to work; was not aware this function was ever used outside services, and yet the old code did do this.

Is it actually necessary to collect info at this level of granularity? Would it be sufficient to just in the Program remember info from file loading to know file references or something?